src/ndb_p1.c

```diff
@@ -0,0 +1,351 @@
+    log_error("Wrong index in ndb_ephe_release_candidate: %d", cand_index);
+    return;
+  }
+  ephe_candidates[cand_index].used = false;
+  log_info("Released slot %d", cand_index);
+}
+
+bool ndb_check_candidate(const ephemeris_t *new, s16 idx) {
+  bool r = memcmp(&ephe_candidates[idx].ephe, new, sizeof(ephemeris_t)) == 0;
+  ndb_ephe_release_candidate(idx);
+  return r;
+}
+
+ndb_ephemeris_status_t ndb_is_ephemeris_reliable(const ephemeris_t *new)
+{
+  ndb_ephemeris_status_t r;
+  ephemeris_t existing;
+  ndb_ephemeris_read(new->sid, &existing);
+  s16 cand_index = ndb_ephe_find_candidate(new);
+
+  chMtxLock(&cand_list_access);
+
+  if (!existing.valid) {
+    log_info_sid(new->sid,"existing ephemeris is invalid or unhealthy");
+    log_info_sid(new->sid,"first reception of ephemeris");
+    if (cand_index != -1) {
+      log_info("Should never get here! %s, %d", __FILE__, __LINE__);
+      ndb_ephe_release_candidate(cand_index);
+    }
+    chMtxUnlock(&cand_list_access);
+    return EPHE_NEW_TRUSTED;
+  }
+
+  /* Ephemeris for this SV was stored to the database already */
+  if (memcmp(&existing, new, sizeof(ephemeris_t)) == 0) {
+    /* New one is identical to the one in DB, no need to do anything */
+    log_info_sid(new->sid, "new ephemeris is identical");
+    if (cand_index != -1)
+      ndb_ephe_release_candidate(cand_index);
+    chMtxUnlock(&cand_list_access);
+    return EPHE_IDENTICAL;
+  }
+
+  if (cand_index != -1) {
+    /* Candidate was added already */
+    if(ndb_check_candidate(new, cand_index)) {
+      log_info_sid(new->sid, "ephemeris candidate trusted");
+      r = EPHE_NEW_TRUSTED;
+    } else {
+      log_info_sid(new->sid, "ephemeris candidate not trusted");
+      r = EPHE_CAND_MISMATCH;
+    }
+  } else {
+    /* New one is not in candidate list yet, try to put it
+     * to an empty slot */
+    log_info_sid(new->sid, "ephemeris candidate added for verification");
+    ndb_ephe_try_adding_candidate(new);
+    r = EPHE_NEW_CANDIDATE;
+  }
+
+  chMtxUnlock(&cand_list_access);
+  return r;
+}
+
+enum ndb_op_code ndb_ephemeris_store(ephemeris_t *e, enum ndb_data_source src)
+{
+  if (!e->valid || !e->healthy) {
+    return NDB_ERR_BAD_PARAM;
+  }
+
+  u16 idx = sid_to_global_index(e->sid);
+
+  switch (ndb_is_ephemeris_reliable(e)) {
+    case EPHE_IDENTICAL:
+      return NDB_ERR_NONE;
+    case EPHE_NEW_TRUSTED:
+      return ndb_update(&ndb_ephemeris[idx], e, sizeof(ephemeris_t), src,
+                        &ndb_ephemeris_md[idx], 0xff, NDB_IE_VALID);
+    case EPHE_NEW_CANDIDATE:
+    case EPHE_CAND_MISMATCH:
+      return NDB_ERR_UNRELIABLE_DATA;
+  }
+  log_info("Should never get here! %s, %d", __FILE__, __LINE__);
+  return NDB_ERR_ALGORITHM_ERROR;
+}
+
+enum ndb_op_code ndb_ephemeris_info(gnss_signal_t sid, u8* v, u8* h,
+                                    gps_time_t* toe, u8* fit_interval)
+{
+  u16 idx = sid_to_global_index(sid);
+  ndb_lock();
+  *v = ndb_ephemeris[idx].valid;
```

Comment:
I vote making it mandatory to provide valid pointers for all parameters. Otherwise `v` and `h` should also be optional.